src/Interpreters/InterpreterInsertQuery.cpp

```diff
@@ -232,6 +234,28 @@ BlockIO InterpreterInsertQuery::execute()
                 out_streams_size = std::min(size_t(settings.max_insert_threads), res.pipeline.getNumStreams());
 
             res.pipeline.resize(out_streams_size);
+
+            /// Allow to insert Nullable into non-Nullable columns, NULL values will be added as defaults values.
+            if (getContext()->getSettingsRef().insert_null_as_default)
+            {
+                auto query_columns_names_and_types = query_sample_block.getNamesAndTypesList();
+
+                const auto & input_columns = res.pipeline.getHeader().getColumns();
+                const auto & query_columns = query_sample_block.getColumns();
+                const auto & output_columns = metadata_snapshot->getColumns();
+
+                size_t col_idx = 0;
+                for (const auto & column : query_columns_names_and_types)
+                {
+                    if (input_columns[col_idx]->isNullable() && !query_columns[col_idx]->isNullable() && output_columns.hasDefault(column.name))
+                    {
+                        auto nullable_column = ColumnNullable::create(query_columns[col_idx], ColumnUInt8::create(query_columns[col_idx]->size(), 0));
+                        ColumnWithTypeAndName new_column(std::move(nullable_column), std::make_shared<DataTypeNullable>(column.type), column.name);
```

Comment:
Note: there are functions `ColumnPtr makeNullable(const ColumnPtr & column);` (in `ColumnNullable.h`) and `DataTypePtr makeNullable(const DataTypePtr & type);` (in `DataTypeNullable.h`)

Reply:
Also, could please add comment why do we need to change `query_sample_block` here?
And why it is safe (because `AddingDefaultBlockOutputStream` will remove nullable)